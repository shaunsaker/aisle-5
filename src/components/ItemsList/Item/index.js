import React from 'react';
import PropTypes from 'prop-types';
import { View, Text } from 'react-native';
import Animator from 'react-native-simple-animators';

import CheckBox from '../../CheckBox';
import ShoppingItem from '../../ShoppingItem';
import Counter from '../../Counter';

import styles from './styles';

const propTypes = {
  name: PropTypes.string,
  isChecked: PropTypes.bool,
  quantity: PropTypes.number,
  handleSetIsChecked: PropTypes.func,
  handleSetQuantity: PropTypes.func,
};

const defaultProps = {};

const Item = ({ name, isChecked, quantity, handleSetIsChecked, handleSetQuantity }) => {
  return (
    <View style={styles.container}>
      <View style={styles.checkBoxContainer}>
        <CheckBox isChecked={isChecked} handlePress={handleSetIsChecked} />
      </View>

      <View style={styles.shoppingItemContainer}>
        <ShoppingItem name={name} quantity={quantity} />
      </View>

      <Animator
        type="translateX"
        initialValue={0}
        finalValue={100}
        shouldAnimateIn={isChecked}
        shouldAnimateOut={!isChecked}
      >
        <Animator
          type="scale"
          initialValue={1}
          finalValue={0}
          shouldAnimateIn={isChecked}
          shouldAnimateOut={!isChecked}
        >
          <Counter value={quantity} handleChange={handleSetQuantity} />
        </Animator>
      </Animator>
    </View>
  );
};

Item.propTypes = propTypes;
Item.defaultProps = defaultProps;

export default Item;
