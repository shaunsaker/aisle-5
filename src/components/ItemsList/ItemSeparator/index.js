import React from 'react';
import { View } from 'react-native';

import styles from './styles';

const propTypes = {};

const defaultProps = {};

const ItemSeparator = () => {
  return <View style={styles.container} />;
};

ItemSeparator.propTypes = propTypes;
ItemSeparator.defaultProps = defaultProps;

export default ItemSeparator;
