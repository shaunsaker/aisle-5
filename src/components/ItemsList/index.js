import React from 'react';
import PropTypes from 'prop-types';
import { FlatList } from 'react-native';
import Animator from 'react-native-simple-animators';

import styleConstants from '../../styleConstants';

import styles from './styles';

import Swipeable from '../Swipeable';
import Item from './Item';

export default class ItemsList extends React.Component {
  constructor(props) {
    super(props);

    this.scrollToEnd = this.scrollToEnd.bind(this);
    this.setDidChange = this.setDidChange.bind(this);
    this.onSwipeStart = this.onSwipeStart.bind(this);
    this.onSwipeEnd = this.onSwipeEnd.bind(this);
    this.setScrollEnabled = this.setScrollEnabled.bind(this);
    this.renderItem = this.renderItem.bind(this);

    this.itemWidth = styleConstants.dimensions.window.width;
    this.itemHeight = 51;
    this.maxItemsVisible = Math.ceil(
      (styleConstants.dimensions.window.height - 65 - 50) / this.itemHeight,
    ); // - headerbar - tabbar

    this.state = {
      didChange: false,
      scrollEnabled: true,
    };
  }

  static propTypes = {
    data: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        isChecked: PropTypes.bool,
      }),
    ),
    handleSetIsChecked: PropTypes.func,
    handleSetQuantity: PropTypes.func,
    handleRemoveItem: PropTypes.func,
  };

  static defaultProps = {};

  componentDidUpdate(prevProps) {
    const { data } = this.props;

    if (data.length !== prevProps.data.length) {
      // We only want to animate new list items in
      this.setDidChange(true);
      this.scrollToEnd();
    }
  }

  scrollToEnd() {
    this.flatList.scrollToEnd();
  }

  setDidChange(didChange) {
    this.setState({ didChange });
  }

  onSwipeStart() {
    this.setScrollEnabled(false);
  }

  onSwipeEnd() {
    this.setScrollEnabled(true);
  }

  setScrollEnabled(scrollEnabled) {
    this.setState({
      scrollEnabled,
    });
  }

  renderItem({ item, index }) {
    const { didChange } = this.state;
    const { handleSetIsChecked, handleSetQuantity, handleRemoveItem } = this.props;

    const shouldAnimate = didChange && index < this.maxItemsVisible;

    return (
      <Swipeable
        width={this.itemWidth}
        onSwipeStart={this.onSwipeStart}
        onSwipeEnd={this.onSwipeEnd}
        onSwiped={() => handleRemoveItem(item.id)}
      >
        <Animator
          type="translateX"
          initialValue={shouldAnimate ? styleConstants.dimensions.window.width : 0}
          finalValue={0}
          shouldAnimateIn
          style={styles.itemContainer}
        >
          <Item
            {...item}
            handleSetIsChecked={() => handleSetIsChecked(item.id, !item.isChecked)}
            handleSetQuantity={(quantity) => handleSetQuantity(item.id, quantity)}
          />
        </Animator>
      </Swipeable>
    );
  }

  render() {
    const { scrollEnabled } = this.state;
    const { data } = this.props;

    return (
      <FlatList
        ref={(c) => {
          this.flatList = c;
        }}
        keyExtractor={({ id }) => id}
        data={data}
        renderItem={this.renderItem}
        style={styles.container}
        contentContainerStyle={styles.contentContainer}
        getItemLayout={(data, index) => ({
          length: this.itemHeight,
          offset: this.itemHeight * index,
          index,
        })}
        scrollEnabled={scrollEnabled}
        bounces={false}
      />
    );
  }
}
