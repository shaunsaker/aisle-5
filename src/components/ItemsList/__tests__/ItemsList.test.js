import React from 'react';
import renderer from 'react-test-renderer';

import ItemsList from '..';
import ITEM from '../../../mockData/item';

jest.mock('../Item', () => 'Item');
jest.mock('../../ItemSeparator', () => 'ItemSeparator');

// Fixes _bezier is not a function bug
jest.useFakeTimers();

describe('ItemsList', () => {
  const spies = [];
  const dispatch = jest.fn();
  const data = [ITEM];
  const newData = [ITEM, { id: '2' }];

  describe('renders', () => {
    it('renders with minimum required props', () => {
      const component = renderer.create(<ItemsList data={data} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    it('should handle scrollToEnd', () => {
      const component = renderer.create(<ItemsList data={data} />);
      const instance = component.getInstance();

      instance.scrollToEnd();
    });

    it('should handle setDidChange', () => {
      const component = renderer.create(<ItemsList data={data} />);
      const instance = component.getInstance();

      expect(instance.state.didChange).toEqual(false);

      instance.setDidChange(true);

      expect(instance.state.didChange).toEqual(true);
    });

    it('should handle onSwipeStart', () => {
      spies[0] = jest.spyOn(ItemsList.prototype, 'setScrollEnabled');
      const component = renderer.create(<ItemsList data={data} />);
      const instance = component.getInstance();

      instance.onSwipeStart();

      expect(spies[0]).toHaveBeenCalledWith(false);
    });

    it('should handle onSwipeEnd', () => {
      spies[0] = jest.spyOn(ItemsList.prototype, 'setScrollEnabled');
      const component = renderer.create(<ItemsList data={data} />);
      const instance = component.getInstance();

      instance.onSwipeEnd();

      expect(spies[0]).toHaveBeenCalledWith(true);
    });

    it('should handle setScrollEnabled', () => {
      const component = renderer.create(<ItemsList data={data} />);
      const instance = component.getInstance();

      expect(instance.state.scrollEnabled).toEqual(true);

      instance.setScrollEnabled(false);

      expect(instance.state.scrollEnabled).toEqual(false);
    });

    it('should handle renderItem', () => {
      // NOTE: We assume this works (handled by RN FlatList)
    });

    it('should handle renderItemSeparator', () => {
      // NOTE: We assume this works (handled by RN FlatList)
    });
  });

  describe('lifecycle methods', () => {
    it('should call methods in componentDidUpdate', () => {
      spies[0] = jest.spyOn(ItemsList.prototype, 'setDidChange');
      spies[1] = jest.spyOn(ItemsList.prototype, 'scrollToEnd');
      const component = renderer.create(<ItemsList data={data} />);

      component.update(<ItemsList data={newData} />);

      expect(spies[0]).toHaveBeenCalledWith(true);
      expect(spies[1]).toHaveBeenCalled();
    });
  });

  // NOTE: We assume the actions are handled correctly

  afterEach(() => {
    spies.forEach((spy) => {
      if (spy) {
        spy.mockClear();
      }
    });
    dispatch.mockClear();
  });
});
