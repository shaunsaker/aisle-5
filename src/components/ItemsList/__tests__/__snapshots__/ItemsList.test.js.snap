// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemsList renders renders with minimum required props 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  bounces={false}
  contentContainerStyle={Object {}}
  data={
    Array [
      Object {
        "id": "1",
        "isChecked": true,
        "name": "Beer",
        "quantity": 6,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEnabled={true}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={Object {}}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <View
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(255, 0, 0, 0)",
              "left": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <Item
              handleSetIsChecked={[Function]}
              handleSetQuantity={[Function]}
              id="1"
              isChecked={true}
              name="Beer"
              quantity={6}
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
